homeassistant/components/izone/climate.py

```diff
@@ -0,0 +1,518 @@
+"""Support for the iZone HVAC."""
+import logging
+from typing import Optional, List
+
+from pizone import Zone, Controller
+
+from homeassistant.core import callback
+from homeassistant.components.climate import ClimateDevice
+from homeassistant.components.climate.const import (
+    HVAC_MODE_HEAT_COOL, HVAC_MODE_COOL, HVAC_MODE_DRY,
+    HVAC_MODE_FAN_ONLY, HVAC_MODE_HEAT, HVAC_MODE_OFF,
+    FAN_LOW, FAN_MEDIUM, FAN_HIGH, FAN_AUTO,
+    PRESET_ECO, PRESET_NONE,
+    SUPPORT_FAN_MODE, SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.const import (
+    ATTR_TEMPERATURE, PRECISION_HALVES, TEMP_CELSIUS)
+from homeassistant.helpers.temperature import display_temp as show_temp
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from .const import (
+    DATA_DISCOVERY_SERVICE, IZONE,
+    DISPATCH_CONTROLLER_DISCOVERED, DISPATCH_CONTROLLER_DISCONNECTED,
+    DISPATCH_CONTROLLER_RECONNECTED, DISPATCH_CONTROLLER_UPDATE,
+    DISPATCH_ZONE_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+_IZONE_FAN_TO_HA = {
+    Controller.Fan.LOW: FAN_LOW,
+    Controller.Fan.MED: FAN_MEDIUM,
+    Controller.Fan.HIGH: FAN_HIGH,
+    Controller.Fan.AUTO: FAN_AUTO,
+}
+
+
+async def async_setup_entry(hass: HomeAssistantType, config: ConfigType,
+                            async_add_entities):
+    """Initialize an IZone Controller."""
+    disco = hass.data[DATA_DISCOVERY_SERVICE]
+
+    @callback
+    def init_controller(controller: Controller):
+        """Register the controller device and the containing zones."""
+        device = ControllerDevice(controller)
+        async_add_entities([device])
+        async_add_entities(device.zones.values())
+        _LOGGER.info("Controller UID=%s added", controller.device_uid)
+
+    # create any components not yet created
+    for controller in disco.controllers.values():
+        init_controller(controller)
+
+    # connect to register any further components
+    async_dispatcher_connect(
+        hass, DISPATCH_CONTROLLER_DISCOVERED,
```

Comment:
We should not connect to the same signal as used internally in `DiscoveryService`.

Reply:
See above.